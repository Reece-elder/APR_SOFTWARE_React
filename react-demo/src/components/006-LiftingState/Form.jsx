const Form = (props) => {

    const {textProp, submitProp, updateText, funcProp} = props;

    return ( 
        <div>
            <h4> Form </h4>
            <form onSubmit={submitProp}>
                <label> Text to display: </label>
                <input type="text" value={textProp} onChange={updateText} name="displayText"/>
                <button type="submit" onClick={submitProp}> Submit! </button>
            </form>
        </div>
     );
}
 
export default Form;