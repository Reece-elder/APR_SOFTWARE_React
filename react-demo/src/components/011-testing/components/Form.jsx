import { useState } from "react";

const Form = () => {

    const [username, setUsername] = useState("");

    const handleSubmit = (e) => {
        e.preventDefault();
        setUsername("");
    }

    return (
        <div>
            <form onSubmit={handleSubmit}>
                <label> username: </label>
                <input id="username" type="text" name="username" value={username} onChange={(event) => setUsername(event.target.value)}/>
                <button id="submitButton" type="submit" onClick={handleSubmit}> Submit Button </button>
            </form>
            <h3> The username is: {username} </h3>
        </div>
    )

}

export default Form;