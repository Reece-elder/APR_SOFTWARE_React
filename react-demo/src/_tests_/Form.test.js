import Form from '../components/011-testing/components/Form'

import { render } from "@testing-library/react";
import {create} from 'react-test-renderer';
import userEvent from "@testing-library/user-event";

describe(`Form testing`,() => {

    describe(`onSubmit test`, () => {

        // Parent level arranging
        const data = "test_name";

        it(`Should update the value of the input (which updates state) when text is typed in`, () => {

            // ? Arrange
            // render(Form) is being deconstructed and saved as 'container'
            const {container} = render(<Form/>);

            const textInput = container.querySelector('#username') // --> ID
            // const textInput = container.querySelector(`[name="username"]`) --> Looking for property / prop
            // const textInput = container.querySelector(`.username`) --> class
            // console.log(textInput);
            expect(textInput.value).toEqual("");

            // ? Act - simulated typing
            userEvent.type(textInput, data);

            // ? Assert
            expect(textInput.value).toEqual(data);

        });

        it(`Should call handleSubmit() when the submit button is clicked`, () => {

            const mockSubmit = jest.fn();

            // Arrange
            const {container} = render(<Form/>);

            const submitButton = container.querySelector('#submitButton');

            // Act
            // submitButton.onClick = mockSubmit();
            userEvent.click(submitButton, mockSubmit())

            // Assert
            expect(mockSubmit).toHaveBeenCalled();

        });

    });

    it(`Should match the snapshot`, () => {
        const instance = create(<Form/>).toJSON();
        expect(instance).toMatchSnapshot();
    })


    
})