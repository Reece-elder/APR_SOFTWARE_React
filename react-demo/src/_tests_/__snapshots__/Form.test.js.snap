// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form testing Should match the snapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
       username: 
    </label>
    <input
      id="username"
      name="username"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      id="submitButton"
      onClick={[Function]}
      type="submit"
    >
       Submit Button 
    </button>
  </form>
  <h3>
     The username is: 
    
     
  </h3>
</div>
`;
